homeassistant/components/weather/met.py

```diff
@@ -0,0 +1,267 @@
+        except (ExpatError, IndexError) as err:
+            try_again(err)
+            return
+
+        async_call_later(self.hass, 60*60, self._fetching_data)
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from Met.no."""
+        if self._weather_data is None:
+            return
+
+        now = dt_util.utcnow()
+
+        ordered_entries = []
+        for time_entry in self._weather_data['product']['time']:
+            valid_from = dt_util.parse_datetime(time_entry['@from'])
+            valid_to = dt_util.parse_datetime(time_entry['@to'])
+
+            if now >= valid_to:
+                # Has already passed. Never select this.
+                continue
+
+            average_dist = (abs((valid_to - now).total_seconds()) +
+                            abs((valid_from - now).total_seconds()))
+
+            ordered_entries.append((average_dist, time_entry))
+
+        if not ordered_entries:
+            return
+        ordered_entries.sort(key=lambda item: item[0])
+
+        self._temperature = get_forecast('temperature', ordered_entries)
+        self._condition = CONDITIONS.get(get_forecast('symbol',
+                                                      ordered_entries))
+        self._pressure = get_forecast('pressure', ordered_entries)
+        self._humidity = get_forecast('humidity', ordered_entries)
+        self._wind_speed = get_forecast('windSpeed', ordered_entries)
+        self._wind_bearing = get_forecast('windDirection', ordered_entries)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def condition(self):
+        """Return the current condition."""
+        return self._condition
+
+    @property
+    def temperature(self):
+        """Return the temperature."""
+        return self._temperature
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def pressure(self):
+        """Return the pressure."""
+        return self._pressure
+
+    @property
+    def humidity(self):
+        """Return the humidity."""
+        return self._humidity
+
+    @property
+    def wind_speed(self):
+        """Return the wind speed."""
+        return self._wind_speed
+
+    @property
+    def wind_bearing(self):
+        """Return the wind direction."""
+        return self._wind_bearing
+
+    @property
+    def attribution(self):
+        """Return the attribution."""
+        return CONF_ATTRIBUTION
+
+
+def get_forecast(param, data):
+    """Retrieve forecast parameter."""
+    try:
+        for (_, selected_time_entry) in data:
+            loc_data = selected_time_entry['location']
+
+            if param not in loc_data:
+                continue
+
+            if param == 'precipitation':
+                new_state = loc_data[param]['@value']
```

Comment:
We should extract this into a third party lib.

Reply:
Ok 👍